design-pattern.md

```diff
@@ -429,3 +429,360 @@ public class Main {
+  @Override
+  public void aggiorna() {
+    System.out.println(
+      "Aggiornamento della 
+      visualizzazione GUI");
+  }
+}
+```
+
+Definiamo la classe `Subject`, in questo esempio la chiamiamo `Database`, che ha un `ArrayList` di `Visualizzazione` e un metodo `aggiornaVisualizzazioni()` che chiama il metodo `aggiorna()` di ogni `Visualizzazione`:
+
+```java
+public class Database {
+  private ArrayList<Visualizzazione> 
+    visualizzazioni;
+
+  public Database() {
+    visualizzazioni = new ArrayList<>();
+  }
+
+  public void aggiungiVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.add(visualizzazione);
+  }
+
+  public void rimuoviVisualizzazione
+  (Visualizzazione visualizzazione) {
+    visualizzazioni.remove(visualizzazione);
+  }
+
+  protected void aggiornaVisualizzazioni() {
+    for(Visualizzazione visualizzazione
+    : visualizzazioni) {
+      visualizzazione.aggiorna();
+    }
+  }
+}
+```
+
+Si noti che nei metodi per aggiungere o rimuovere una `Visualizzazione`, abbiamo l'interfaccia `Visualizzazione` come tipo di parametro. Questo ci permette di aggiungere dinamicamente qualsiasi tipo di `Visualizzazione` che implementa l'interfaccia `Visualizzazione`.
+
+---
+
+<!-- _class: due -->
+
+### State
+
+**State** è un pattern che permette di cambiare il comportamento di un oggetto in base allo stato in cui si trova. È molto simile a **Strategy**.
+
+#### UML
+
+![](./immagini/state.svg)
+
+In questo diagramma, troviamo `State` che è l'interfaccia che definisce uno stato _generico_, `ConcreteState` che estende `State` e definisce il comportamento in quello specifico stato e `Context` che è l'oggetto che cambia stato.
+
+#### Esempio
+
+Supponiamo di voler implementare il TCP. La sequenza di azione da compiere è rappresentabile con un automa a stati finiti. La classe TCPConnection deve gestire l'apertura della connessione, la chiusura e le varie fasi intermedie. Usiamo uno **State**.
+
+Definiamo la classe astratta `State`, in questo esempio la chiamiamo `TCPState`, che definisce i metodi `open()`, `close()`, `acknowledge()` e `transition()`:
+
+```java
+public abstract class TCPState {
+  protected TCPConnection connection;
+
+  public TCPState(TCPConnection connection) {
+    this.connection = connection;
+  }
+
+  public abstract void open();
+  public abstract void close();
+  public abstract void acknowledge();
+
+  public void transition(TCPState state) {
+    connection.setState(state);
+  }
+}
+```
+
+Definiamo le varie classi `ConcreteState`, in questo esempio le chiamiamo `TCPListen`, ecc., che estendono `TCPState`:
+
+```java
+public class TCPListen extends TCPState {
+  public TCPListen(TCPConnection connection) {
+    super(connection);
+  }
+
+  @Override
+  public void open() {
+    // open connection
+    this.transition(new TCPEstablished(connection));
+  }
+  @Override
+  public void close() { ... }
+  @Override
+  public void acknowledge() { ... }
+}
+```
```

Comment:
Osserviamo la differenza **fondamentale** tra _Strategy_ e _State_. Alla fine dell'esecuzione del metodo open, avviene una transizione dello stato da `TCPListen` a `TCPEstablished`. Infatti sia in _Strategy_ che in _State_ il comportamento della classe cambia a runtime, la differenza tra i due è che nel primo caso è l'utente che decide quando cambiare il comportamento, nel secondo caso il comportamento cambia perchè avviene una transizione di stato che è completamente trasparente all'utente della classe e dipende solo dallo stato di partenza e dal metodo invocato.

(Lo aggiungo io dopo il merge)